File: leetcode/python/000242.valid-anagram.py
class Solution:
    # Time complexity: O(n)
    # Space complexity: O(n)
    def isAnagram(self, s: str, t: str) -> bool:
        s_char_map = {}
        t_char_map = {}

        for ch in s:
            s_char_map[ch] = s_char_map.get(ch, 0) + 1

        for ch in t:
            t_char_map[ch] = t_char_map.get(ch, 0) + 1

        return s_char_map == t_char_map

    # Time complexity: O(n)
    # Space complexity: O(n)
    def isAnagram2(self, s: str, t: str) -> bool:
        map = {}

        for ch in s:
            map[ch] = map.get(ch, 0) + 1

        for ch in t:
            map[ch] = map.get(ch, 0) - 1

        for v in map.values():
            if v != 0:
                return False

        return True

File: leetcode/javascript/000242.valid-anagram.js
/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 * Time complexity: O(n)
 * Space complexity: O(n)
 */
var isAnagram = function (s, t) {
  const counter = new Map();

  for (const ch of s) {
    const cur = counter.get(ch) ?? 0;
    counter.set(ch, cur + 1);
  }

  for (const ch of t) {
    const cur = counter.get(ch) ?? 0;
    counter.set(ch, cur - 1);
  }

  for (const val of counter.values()) {
    if (val !== 0) return false;
  }

  return true;
};
